src/ucp/tag/offload.c

```diff
@@ -513,26 +513,37 @@ ucp_do_tag_offload_zcopy(uct_pending_req_t *self, uint64_t imm_data,
                                     req->send.msg_proto.tag.tag,
                                     imm_data, iov, iovcnt, 0,
                                     &req->send.state.uct_comp);
-    if (status == UCS_OK) {
-        complete(req, UCS_OK);
-    } else if (status == UCS_INPROGRESS) {
+    if (ucs_unlikely(UCS_STATUS_IS_ERR(status))) {
```

Comment:
can be:
```
if (status == UCS_OK) {
    complete(req, UCS_OK);
} else {
    ucp_request_send_state_advance(req, &state,
                                   UCP_REQUEST_SEND_PROTO_ZCOPY_AM,
                                   status);
    if (status == UCS_INPROGRESS) {
        status = UCS_OK;
    }
}

return status;
```


or even wrap to the separate function to not duplicate the code for offload/normal Zcopy's implementations

Reply:
better to handle 3 cases separately since this way inline `ucp_request_send_state_advance` can be optimized for IN_PROGRESS case

Reply:
WRT
> or even wrap to the separate function to not duplicate the code for offload/normal Zcopy's implementations

good idea, will add common func to proto_am.inl

Reply:
> better to handle 3 cases separately since this way inline `ucp_request_send_state_advance` can be optimized for IN_PROGRESS case

I see